File: negative_marking/287_find_the_duplicate_number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        # Negative marking
        for i in range(len(nums)):
            pos = abs(nums[i]) - 1
            if nums[pos] < 0:
                return abs(nums[i])
            nums[pos] *= -1
        return -1

        # Binary Search
        # def binSearch(arr, target):
        #     l, r = 0, len(arr)
        #     while l <= r:
        #         m = (l + r) >> 1
        #         if arr[m] == target:
        #             return target
        #         elif arr[m] > target:
        #             r = m - 1
        #         else:
        #             l = m + 1
        #     return -1

        # nums.sort()
        # n = len(nums)

        # for i in range(n):
        #     num = nums[i]
        #     if binSearch(nums[i+1:], num) != -1:
        #         return num
            
        # return -1


        # Brute-force
        # n = len(nums)

        # for i in range(n):
        #     for j in range(i+1, n):
        #         if nums[j] == nums[i]:
        #             return nums[i]
        # return -1


        

File: array/287_find_the_duplicate_number.java
class Solution {
    public int findDuplicate(int[] nums) {
        for (int num : nums) {
            int pos = Math.abs(num);

            if (nums[pos - 1] < 0) return pos;
            nums[pos - 1] *= -1;
        }
        return -1;

        // int n = nums.length;
        // boolean[] seen = new boolean[n+1];

        // for (int num : nums) {
        //     if (seen[num - 1]) return num;
        //     seen[num - 1] = true;
        // }
        // return -1;
    }
}